File: 백준/Gold/1504. 특정한 최단 경로/특정한 최단 경로.cc
#include <iostream>
#include <queue>
#include <vector>
#include <climits>

using namespace std;

int n, e, a, b, c, v1, v2;
int dist[801];

void dijkstra(int node, vector<vector<pair<int ,int>>> &graph) {
	priority_queue<pair<int, int>> pq;
	fill(dist, dist+n+1, INT_MAX);
	dist[node] = 0;
	pq.push({0, node});

	while (!pq.empty()) {
		int current_dist = -pq.top().first;
		int currnet_node = pq.top().second;
		pq.pop();

		for (int i = 0; i < graph[currnet_node].size(); i++) {
			int next_node = graph[currnet_node][i].first;
			int next_dist = current_dist + graph[currnet_node][i].second;

			if (next_dist < dist[next_node]) {
				dist[next_node] = next_dist;
				pq.push({-next_dist, next_node});
			}
		}
	}
}

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);
    cout.tie(NULL);

    cin >> n >> e;
    vector<vector<pair<int ,int>>> graph(n+1);
    for (int i = 0; i < e; i++) {
        cin >> a >> b >> c;
        graph[a].push_back({b, c});
        graph[b].push_back({a, c});
    }
    cin >> v1 >> v2;

    long long route1 = 0, route2 = 0;
    
    // 첫 번째 경로: 1 -> v1 -> v2 -> N
    dijkstra(1, graph);
    long long dist_1_to_v1 = dist[v1];
    
    dijkstra(v1, graph);
    long long dist_v1_to_v2 = dist[v2];
    
    dijkstra(v2, graph);
    long long dist_v2_to_n = dist[n];
    
    route1 = (dist_1_to_v1 < INT_MAX && dist_v1_to_v2 < INT_MAX && dist_v2_to_n < INT_MAX) 
             ? dist_1_to_v1 + dist_v1_to_v2 + dist_v2_to_n 
             : INT_MAX;

    // 두 번째 경로: 1 -> v2 -> v1 -> N
    dijkstra(1, graph);
    long long dist_1_to_v2 = dist[v2];
    
    dijkstra(v2, graph);
    long long dist_v2_to_v1 = dist[v1];
    
    dijkstra(v1, graph);
    long long dist_v1_to_n = dist[n];
    
    route2 = (dist_1_to_v2 < INT_MAX && dist_v2_to_v1 < INT_MAX && dist_v1_to_n < INT_MAX) 
             ? dist_1_to_v2 + dist_v2_to_v1 + dist_v1_to_n 
             : INT_MAX;

    if (min(route1, route2) == INT_MAX) cout << -1;
    else cout << min(route1, route2);
}

File: 백준/Gold/1504. 특정한 최단 경로/특정한 최단 경로.java
import java.io.*;
import java.util.*;

public class Main {
    static final long MAX_DISTANCE = 800L * 200_000 * 1_000 + 1;

    static long answer;
    static int n, e, v1, v2;
    static ArrayList<ArrayList<int[]>> graph;

    static class Node {
        int node;
        long dist;

        Node(int node, long dist) {
            this.node = node;
            this.dist = dist;
        }
    }

    private static long dijkstra(int start, int end) {
        long[] res = new long[n+1];
        Arrays.fill(res, MAX_DISTANCE);
        res[start] = 0;

        PriorityQueue<Node> pq = new PriorityQueue<>((n1, n2) -> Long.compare(n1.dist, n2.dist));
        pq.offer(new Node(start, 0));

        while(!pq.isEmpty()) {
            Node cur = pq.poll();

            int node = cur.node;
            long dist = cur.dist;

            if (res[node] < dist) continue;

            for (int info[] : graph.get(node)) {
                int nextNode = info[0];
                int nextDist = info[1];

                if (res[nextNode] > res[node] + nextDist) {
                    res[nextNode] = res[node] + nextDist;
                    pq.offer(new Node(nextNode, res[nextNode]));
                }
            }
        }

        return res[end];
    }

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());

        n = Integer.parseInt(st.nextToken());
        e = Integer.parseInt(st.nextToken());

        graph = new ArrayList<>();
        for (int i = 0; i <= n; i++) graph.add(new ArrayList<>());

        for (int i = 0; i < e; i++) {
            int a, b, c;
            st = new StringTokenizer(br.readLine());

            a = Integer.parseInt(st.nextToken());
            b = Integer.parseInt(st.nextToken());
            c = Integer.parseInt(st.nextToken());

            graph.get(a).add(new int[] {b, c});
            graph.get(b).add(new int[] {a, c});
        }

        st = new StringTokenizer(br.readLine());

        v1 = Integer.parseInt(st.nextToken());
        v2 = Integer.parseInt(st.nextToken());

        long oneToV1 = dijkstra(1, v1);
        long oneToV2 = dijkstra(1, v2);

        long v1ToV2 = dijkstra(v1, v2);

        long v1ToN = dijkstra(v1, n);
        long v2ToN = dijkstra(v2, n);

        long first = oneToV1 + v1ToV2 + v2ToN;
        long second = oneToV2 + v1ToV2 + v1ToN;

        if (first >= MAX_DISTANCE && second >= MAX_DISTANCE) answer = -1;
        else answer = Math.min(first, second);
        
        System.out.println(answer);
    }
}